### ISSUE TITLE
After upgrade from 0.7.5 to 0.7.8 DPOTrainer get RuntimeError: expected scalar type Float but found BFloat16

### ISSUE THREAD

**Issue by Andcircle** [huggingface/trl#1202]
After upgrade from 0.7.5 to 0.7.8 DPOTrainer get RuntimeError: expected scalar type Float but found BFloat16

I'm using KTO loss, haven't tried other loss yet

```
Traceback (most recent call last):
  File "cpt-hf/train/train_dpo.py", line 307, in <module>
    trainer.train()
  File "/home/user/.local/lib/python3.8/site-packages/transformers/trainer.py", line 1537, in train
    return inner_training_loop(
  File "/home/user/.local/lib/python3.8/site-packages/transformers/trainer.py", line 1914, in _inner_training_loop
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/home/user/.local/lib/python3.8/site-packages/transformers/trainer.py", line 2268, in _maybe_log_save_evaluate
    metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
  File "/home/user/.local/lib/python3.8/site-packages/transformers/trainer.py", line 3019, in evaluate
    output = eval_loop(
  File "/home/user/.local/lib/python3.8/site-packages/trl/trainer/dpo_trainer.py", line 1163, in evaluation_loop
    policy_output_decoded, ref_output_decoded = self.get_batch_samples(self.model, random_batch)
  File "/home/user/.local/lib/python3.8/site-packages/trl/trainer/dpo_trainer.py", line 1056, in get_batch_samples
    policy_output = model.generate(
  File "/usr/local/lib/python3.8/dist-packages/peft/peft_model.py", line 988, in generate
    outputs = self.base_model.generate(**kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/user/.local/lib/python3.8/site-packages/transformers/generation/utils.py", line 1764, in generate
    return self.sample(
  File "/home/user/.local/lib/python3.8/site-packages/transformers/generation/utils.py", line 2861, in sample
    outputs = self(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/home/user/.local/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py", line 1199, in forward
    logits = self.lm_head(hidden_states)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: expected scalar type Float but found BFloat16
```

**Comment by younesbelkada** [comment]
Thanks for reporting! can you share a small snippet to illustrate how you encounter the bug ? cc @kashif

**Comment by younesbelkada** [comment]
Also can you confirm you don't get the error with trl==0.7.7 ?

**Comment by younesbelkada** [comment]
Might be related to: https://github.com/huggingface/trl/pull/1143

**Comment by Andcircle** [comment]
@younesbelkada 
here's the code 
```
def create_and_prepare_model():
    compute_dtype = getattr(torch, "float16")

    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True,
        # load_in_8bit=True,
        bnb_4bit_quant_type="nf4",
        bnb_4bit_compute_dtype=compute_dtype,
        bnb_4bit_use_double_quant=True,
    )

    model = AutoModelForCausalLM.from_pretrained(
        model_name, quantization_config=bnb_config, trust_remote_code=True, 
        # device_map='auto',
        device_map={'':torch.cuda.current_device()},
        use_flash_attention_2=True
        )
    print(model)
    
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj"]
        
    peft_config = LoraConfig(
        lora_alpha=alpha,
        lora_dropout=0.1,
        r=rank,
        bias="none",
        task_type="CAUSAL_LM",
        target_modules=target_modules
    )

    tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
    tokenizer.pad_token = tokenizer.eos_token
    tokenizer.padding_side = 'right'

    return model, peft_config, tokenizer


training_arguments = TrainingArguments(
    output_dir=os.path.join(save_dir, run_name),
    per_device_train_batch_size=batch_size,
    gradient_accumulation_steps=accumlate_steps,
    optim="paged_adamw_32bit",
    
    # Can't resume training: Error invalid argument at line 393 in file /mmfs1/gscratch/zlab/timdettmers/git/bitsandbytes/csrc/pythonInterface.c 
    # optim="paged_adamw_8bit",
    
    save_steps=500,
    logging_steps=10,
    learning_rate=lr,
    fp16=True,
    max_grad_norm=0.3,
    num_train_epochs=100,
    warmup_ratio=0.03,
    # group_by_length=True,
    lr_scheduler_type="constant",
    run_name=run_name,
    evaluation_strategy="steps",
    eval_steps=200,
    ddp_find_unused_parameters=False,
    gradient_checkpointing=True,
    # weight_decay=0.01,
    # dataloader_num_workers=NUM_PROC//2
)

model, peft_config, tokenizer = create_and_prepare_model()
model.config.use_cache = False # because of gradient checkpointing


trainer = SFTTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    peft_config=peft_config,
    dataset_text_field="text",
    max_seq_length=length,
    tokenizer=tokenizer,
    args=training_arguments,
    packing=True,
)

trainer.train()
```

**Comment by younesbelkada** [comment]
are you sure this is the correct script? In traceback I see lines from the DPOTrainer

**Comment by Andcircle** [comment]
@younesbelkada sorry I just changed to the correct one, also tried 0.7.7 works fine

**Comment by younesbelkada** [comment]
Thanks @Andcircle I managed to reproduce and will provide a fix soon

**Comment by Andcircle** [comment]
@younesbelkada thanks for the super fast response

**Comment by younesbelkada** [comment]
@Andcircle things are fixed now on Pypi, you can use trl==0.7.9 without any problem
